spanner/benchmark/ycsb.py

```diff
@@ -0,0 +1,259 @@
+
+  # To make a package so it can work with PerfKitBenchmarker.
+  $ cd spanner; tar -cvzf ycsb-python.0.0.5.tar.gz benchmark/*
+
+"""
+
+import numpy
+import optparse
+import random
+import string
+import threading
+import timeit
+
+from google.cloud import spanner
+
+
+KEYS = []
+OPERATIONS = ['readproportion', 'updateproportion', 'scanproportion',
+              'insertproportion']
+
+
+def ParseOptions():
+  """Parses options."""
+  parser = optparse.OptionParser()
+  parser.add_option('-P', '--workload', action='store', dest='workload',
+                    default='', help='The path to a YCSB workload file.')
+  parser.add_option('-p', '--parameter', action='append', dest='parameters',
+                    default=[], help='The key=value pair of parameter.')
+  parser.add_option('-b', '--num_bucket', action='store', type='int',
+                    dest='num_bucket', default=1000,
+                    help='The number of buckets in output.')
+
+  options, args = parser.parse_args()
+
+  parameters = {}
+  parameters['command'] = args[0]
+  parameters['num_bucket'] = options.num_bucket
+
+  for parameter in options.parameters:
+    parts = parameter.strip().split('=')
+    parameters[parts[0]] = parts[1]
+
+  with open(options.workload, 'r') as f:
+    for line in f.readlines():
+      parts = line.split('=')
+      key = parts[0].strip()
+      if key in OPERATIONS:
+        parameters[key] = parts[1].strip()
+
+  return parameters
+
+
+def OpenDatabase(parameters):
+  """Opens a database specified by the parameters from ParseOptions()."""
+  spanner_client = spanner.Client()
+  instance_id = parameters['cloudspanner.instance']
+  instance = spanner_client.instance(instance_id)
+  database_id = parameters['cloudspanner.database']
+  pool = spanner.BurstyPool(int(parameters['num_worker']))
+  database = instance.database(database_id, pool=pool)
+
+  return database
+
+
+def LoadKeys(database, parameters):
+  """Loads keys from database."""
+  global KEYS
+  KEYS = []
+  results = database.execute_sql('SELECT u.id FROM %s u' % parameters['table'])
+
+  for row in results:
+    KEYS.append(row[0])
+
+
+def Read(database, table, key):
+  """Does a single read operation."""
+  with database.snapshot() as snapshot:
+    result = snapshot.execute_sql('SELECT u.* FROM %s u WHERE u.id="%s"' %
+                                  (table, key))
+    for row in result:
+      key = row[0]
+      for i in range(10):
+        field = row[i + 1]
+
+
+def Update(database, table, key):
+  """Does a single update operation."""
+  field = random.randrange(10)
+  value = ''.join(random.choice(string.printable) for i in range(100))
+  with database.batch() as batch:
+    batch.update(table=table, columns=('id', 'field%d' % field),
+                 values=[(key, value)])
+
+
+def Insert(database, table, key):
+  """Does a single insert operation."""
+  raise Exception('Insert is not implemented.')
```

Comment:
Raise NotImplementedError?  

https://softwareengineering.stackexchange.com/questions/231397/is-it-conventional-to-raise-a-notimplementederror-for-methods-whose-implementati

Reply:
Done in my next commit.